### PR TITLE
Vcpkg manifest and binary caching

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/3297

# Binary Cache and Manifest

This PR updates the current CI gates `testing model` to use VCPKG binary cache and manifest mode.

### vcpkg json

Remove the list of vcpkg dependencies for each CI gate.  Instead, a `vcpkg.json` is located at the root of each project and it defines the list of libraries to be installed as a dependency for the project.  This change moves the `responsibility` of defining the vcpkg dependencies from the `CI-configuration` to the project itself.  In the future, when adding a new VCPKG dependency, the user will only need to update the `vcpkg.json` file without updating the ci-configuration.

### vcpkg cache

Remove the step `vcpkg cache` from the ci gates. Instead of manually creating a hash key based on the vcpkg dependencies and using Azure DevOps as a vcpkg cache, this PR removes all this logic and simplify things by using the binary-cache provided by VCPKG. Now it is VCPKG the one in charge of checking if the required binaries are already in the cache and pulling from it to restore. Public pipelines are updated to `only allow reading` from the cache to restore binaries, while `internal` pipelines are are instrumented to be able to `write` binaries to cache. 

In the future, when adding a new vckpg dependency, if the team needs to persist the new library to the cache, the team needs to trigger the internal pipelines within the PR to let them cache the binaries. Otherwise, the cache would be created after merging the PR, when the internal pipelines run automatically.

## Installing vcpkg dependencies

Removed the ci - step that install VCPKG dependencies after cloning vcpkg. Instead, vcpkg is only cloned and set it up to work with the azure sdk repo (toolchain file is set). Then, the `vcpkg.json` file turns on `manifest mode`, which will ask CMake to run `vcpkg install dependencies` before it generates the project.

### Related issues:

https://github.com/Azure/azure-sdk-for-cpp/issues/3087
https://github.com/Azure/azure-sdk-for-cpp/issues/2617
https://github.com/Azure/azure-sdk-for-cpp/issues/2167
https://github.com/Azure/azure-sdk-for-cpp/issues/1765
https://github.com/Azure/azure-sdk-for-cpp/issues/917

### CODE REVIEW

File: sdk/identity/vcpkg.json
Comment (danieljurek) [inline sdk/identity/vcpkg.json:5]: Is this entry required?

**Review by danieljurek — COMMENTED** [review]
Couple of suggestions and some EngSys work on my side.

**Comment by danieljurek** [comment]
Another question: If the SAS key becomes invalid or expires does vcpkg still function or will it fail and break our pipelines?

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (danieljurek) [reply eng/pipelines/templates/steps/vcpkg.yml:19]: * Variable group created ✅
* PR to associate variable group with pipeline 🔃 https://github.com/Azure/azure-sdk-tools/pull/2711 

Use the secret `$(cpp-vcpkg-cache-write-sas)` in the code

File: vcpkg.json
Comment (vhvb1989) [inline vcpkg.json:9]:
```suggestion
}

```

File: sdk/keyvault/azure-security-keyvault-secrets/vcpkg.json
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-secrets/vcpkg.json:8]:
```suggestion
}

```

File: sdk/keyvault/azure-security-keyvault-certificates/vcpkg.json
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/vcpkg.json:8]:
```suggestion
}

```

File: sdk/core/azure-core/vcpkg.json
Comment (vhvb1989) [inline sdk/core/azure-core/vcpkg.json:8]:
```suggestion
}

```

**Comment by weshaggard** [comment]
@vhvb1989 can I get a little more context on this change? Is there an associated issue I can read?

**Comment by vhvb1989** [comment]
> @vhvb1989 can I get a little more context on this change? Is there an associated issue I can read?

shared email with details.  I will create an issue (why not, right :)

**Comment by vhvb1989** [comment]
@weshaggard  I found some issues related to this PR which will be fixed :).  See below


Related issues:
https://github.com/Azure/azure-sdk-for-cpp/issues/3087
https://github.com/Azure/azure-sdk-for-cpp/issues/2617
https://github.com/Azure/azure-sdk-for-cpp/issues/2167
https://github.com/Azure/azure-sdk-for-cpp/issues/1765
https://github.com/Azure/azure-sdk-for-cpp/issues/917

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (gearama) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:80]: can you add a note about what this read/readwrite mean ?

File: eng/pipelines/templates/jobs/ci.tests.yml
Comment (gearama) [inline eng/pipelines/templates/jobs/ci.tests.yml:49]: from where ?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (gearama) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:37]: from where ?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (gearama) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:70]: here you say read

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (gearama) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:38]: why you don't set cache mode here ? but you do earlier

**Review by gearama — COMMENTED** [review]
Overall, we(I) need a PR description to figure out what it's happening here. 
I have a suspicion, alas my level 0 necromancy skills are not advanced enough to allow me to summon the spirit of Sherlock Holmes to perform an investigation on what this will do and how it works.

**Comment by vhvb1989** [comment]
> Overall, we(I) need a PR description to figure out what it's happening here. I have a suspicion, alas my level 0 necromancy skills are not advanced enough to allow me to summon the spirit of Sherlock Holmes to perform an investigation on what this will do and how it works.

I have added PR description listing the main changes/improvements for this PR

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:37]: binary cache mode has 2 operation modes, `read` and `readwrite`.

read will only try to fetch and restore binaries from the cache, while readwrite will push binaries to the cache after building them.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:70]: yes, since this is public pipeline, we only allow read mode.
And in case you wonder, if anyone feels smart and change this to `readwrite` and run the pipeline; the pipeline will ask vcpkg to write built binaries, but VCPKG will fail to push the binaries because only the `internal` pipelines are instrumented with writting access.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:38]: Because the mode is set in the initial ci configuration and it becomes an env var that is consumed by the vcpkg step. The rest of the ci configuration don't care about it.

File: eng/pipelines/templates/jobs/ci.tests.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/ci.tests.yml:49]: from the cache. ( it is set in the vcpkg step, I will add a comment over there for you)

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (vhvb1989) [inline eng/pipelines/templates/steps/vcpkg.yml:16]: @gearama , this is the source from where the pipeline would either read or readwrite

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:19]: We can call it `Daniel's magic`  😄

**Comment by danieljurek** [comment]
> /azp run cpp - core

This should not result in a write to the cache as the secret is not yet available to the pipeline. That PR is here: https://github.com/Azure/azure-sdk-tools/pull/2711

That PR should be approved by folks from the EngSys team, merged, and the relevant automation run before the secret is available to the pipeline.

File: sdk/core/azure-core/vcpkg.json
Comment (antkmsft) [inline sdk/core/azure-core/vcpkg.json:1]: This a bit overlaps with the `vcpkg/vcpkg.json`. Is it possible to integrate them together more tightly / reuse the same file?

If not, is it possible to drop `name` and `version` from these files?

File: sdk/core/azure-core/vcpkg.json
Comment (antkmsft) [inline sdk/core/azure-core/vcpkg.json:6]: Our vcpkg manifest template files (the ones in `./vcpkg/`, do have platform-specific dependencies, for example, Core does not have the openssl dependency on Windows.
```
    {
      "name": "openssl",
      "platform": "!windows & !uwp"
    },
```

File: sdk/core/azure-core/vcpkg.json
Comment (antkmsft) [inline sdk/core/azure-core/vcpkg.json:3]: I see versions are `"1.0.0 + version-string instead of version-semver"` everywhere. Is this intentional? Just to have something so that vcpkg does not complain? Can we remove it?

**Review by antkmsft — COMMENTED** [review]
LGTM, assuming everything works as intended and nothing is broken. Not sure if this is much simpler than what we had, but seems to be more clean way, plus vcpkg team recommended it, let's give it a try.

I had some feedback below.

Plus, for the intra-lib dependencies inside SDK, we use the following, to distinguish between the case when the library is being built as vcpkg vs whole SDK is being built:
```
elseif(NOT AZ_ALL_LIBRARIES)
  find_package(azure-core-cpp "1.3.1" CONFIG QUIET)
  if(NOT azure-core-cpp_FOUND)
    find_package(azure-core-cpp "1.3.1" REQUIRED)
  endif()
endif()
```

Can you verify if it will not go inside `find_package(azure-core-cpp)`? Maybe you can temporarily insert a `message()` in the cmakelists to check. Because if it does go inside the `if`, then it may actually compile against the latest published version, instead of the source tree version. Unless, manifest mode magic makes it to work, in which case we may not need the `NOT AZ_ALL_LIBRARIES` check. In case it does not go inside the `if`, I guess no action is needed, this aspect works just as before. Or maybe we can drop the `NOT AZ_ALL_LIBRARIES` check, and vcpkg will make `find_package(` to work in a smart way?

Also, I noticed there is now another build mode, that George added, with fetching source dependencies. Are you not breaking it, are they still both going to be compatible? We do not test the fetch-build in CI, do we? If not, can you run it, maybe with George, to make sure that "anti-vcpkg" build mode keeps to work as well?

File: sdk/core/azure-core/vcpkg.json
Comment (vhvb1989) [reply sdk/core/azure-core/vcpkg.json:1]: looks like we can use VCPKG_MANIFEST_DIR option and re-use the same vcpkg.json we already have :)

Comment (vhvb1989) [reply sdk/core/azure-core/vcpkg.json:1]: the problem is that our current vcpkg.json is only a template that requires some pre-proccesing before it is useful. We remove the header lines and set the version value during release :)

**Review by gearama — APPROVED** [review]
LGTM

File: sdk/core/azure-core/vcpkg.json
Comment (antkmsft) [reply sdk/core/azure-core/vcpkg.json:1]: @vhvb1989, vcpkg would ignore fields that start with `$` (https://vcpkg.readthedocs.io/en/latest/specifications/manifests/). So, if they allow manifests with no version, we can rename `"version-semver"` to `"$version-semver"`, and then rename it back during the install step: see `az_vcpkg_portfile_prep` cmake macro - it drops the copyright comment, we'll just do the similar thing with the `"$version-semver"`.

Comment (antkmsft) [reply sdk/core/azure-core/vcpkg.json:1]: or maybe, they won't complain at all, that our `version-semver` looks like `"@AZ_LIBRARY_VERSION@"`, then it's even easier - nothing needs to be done.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:141]: Why are you setting this explicitly here? Should it be handled by the template parameter? Also I assume readwrite means that you will need to be internal and have access to secrets to write.

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:16]: Is VCPKG_BINARY_SOURCES a well-known variable that VCPKG picks up?

Comment (weshaggard) [reply eng/pipelines/templates/steps/vcpkg.yml:16]: Is the SAS token needed in read mode as well? If folks want to use this locally would they need to generate a SAS?

Comment (weshaggard) [reply eng/pipelines/templates/steps/vcpkg.yml:16]: Also just so it is clear the SAS will be empty for any public builds as they cannot access that value.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:141]: the way I see it, each service has a `ci.yaml` as the starting point.
the `ci.yaml` invokes `stages/archetype-sdk-client.yml` for both, `internal` and `public`.

Then, `stages/archetype-sdk-client.yml` uses `jobs/archetype-sdk-client.yml`, for both `internal` and `public`. It uses the stage name `Build`.  So, `build` runs public and internal.  Here, when invoking the `build` stage, I am checking `variables['System.TeamProject']` to either set `VcpkgCacheMode: 'readwrite'` or leave it with default value (read).  This covers the vcpkg cache for the build stage only, for both Internal and public.

Then, for internal pipelines, `stages/archetype-sdk-client.yml` also add stage `LiveTest` (jobs/archetype-sdk-tests.yml). Since this is **only** internal, I set `readwrite` directly as a variable in the matrix definition so it becomes an env var.  We don't need to make a input parameter for the stage here because it is always for internal only.

Finally, the `build` stage has one extra job ( out of the generated matrix ) called `GenerateReleaseArtifacts` which defines its own pool and variables.  So I also need to set the cache mode to the `${{ parameters.VcpkgCacheMode }}` so it only reads on public and read/write on internal pipeline.

As an alternative, I could use `readwrite` for everything, and let the public pipelines try to push cache to VCPKG and fail as the SAS secret won't be set.  VCPKG will just write to console something like `failed to push binary to the servers`, but woudn't make the pipeline to fail.  But this approach seems a little lazy, and someone would think that it is actually an error over there.

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [reply eng/pipelines/templates/steps/vcpkg.yml:16]: Reading a little more at https://vcpkg.io/en/docs/users/binarycaching.html is there reason we chose blob storage over say DevOps Artifact feed or something? I mostly ask because they mark the blob support as expermential.

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:11]: I was just making sure that the value is displayed as `Cache SAS: *****`   on private pipelines and empty `Cache SAS: ` on public pipelines.  But I will remove it.  I wanted to be sure the secret was not showing up

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:16]: > Is VCPKG_BINARY_SOURCES a well-known variable that VCPKG picks up?

Yes. https://vcpkg.io/en/docs/users/binarycaching.html

> Is the SAS token needed in read mode as well? If folks want to use this locally would they need to generate a SAS?

No, public pipelines, and anyone locally can just set `VCPKG_BINARY_SOURCES=clear;x-azblob,https://cppvcpkgcache.blob.core.windows.net/public-vcpkg-container,,read` to ask VCPKG to pull from there.  

> Also just so it is clear the SAS will be empty for any public builds as they cannot access that value.

Yes, that's the desired behavior :)

Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:15]: > Reading a little more at https://vcpkg.io/en/docs/users/binarycaching.html is there reason we chose blob storage over say DevOps Artifact feed or something? I mostly ask because they mark the blob support as experimental.

DevOps and GitHub require Nuget.  We would need to install `Nuget + Mono` for non-windows machines. I didn't like that.
I talked to VCPKG team and confirm that they will only keep adding features to the Azure Storage source ( like auth modes, config, etc ), but it won't be removed.  So I went that route as it won't require `mono` or anything else

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:141]: Should we just remove the notion of the mode everywhere and simply check if we have a SAS token then set mode to readwrite and if not just read.

Then we ifdef the access of the SAS token on if the TeamProject is internal.

Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:141]: Yes, I like the idea! :)

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:19]: You should put an if around this to only try and access this sas token for internal builds.

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:8]: Without the if I suggested in the other comment this will always exist as a variable and thus this Test-Path will always return true.

Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:8]: ah, I made a test locally and Test-Path returns false if the environment variable is set with empty string.  But yes, let me try what you suggest.

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:21]: You should probably at least check that VCPKG_CACHE_SAS has a value and if not give an error message as that might help diagnostics if something is not setup correctly.

**Review by weshaggard — APPROVED** [review]
Left one other minor clean-up comment but otherwise looks good.

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:19]: https://github.com/Azure/azure-sdk-for-cpp/issues/3418

File: sdk/core/azure-core/vcpkg.json
Comment (vhvb1989) [reply sdk/core/azure-core/vcpkg.json]: Can't remove it. it is required

File: sdk/core/azure-core/vcpkg.json
Comment (vhvb1989) [reply sdk/core/azure-core/vcpkg.json]: used it

File: sdk/core/azure-core/vcpkg.json
Comment (vhvb1989) [reply sdk/core/azure-core/vcpkg.json:1]: The next step would be an automatic procedure to sync the files

**Comment by danieljurek** [comment]
There's a timeout in `cpp - keyvault - ci` ... Rekicking to see if we can get it to green -- https://dev.azure.com/azure-sdk/public/_build/results?buildId=1422059&view=results

**Comment by vhvb1989** [comment]
> There's a timeout in `cpp - keyvault - ci` ... Rekicking to see if we can get it to green -- https://dev.azure.com/azure-sdk/public/_build/results?buildId=1422059&view=results

Thank you.
It might be that these generate-tests are installing a `vcpkg` package that is never installed with the internal pipeline.  Hence, the cache is never created.  
I will add this generate stage to the internal pipeline.

**Comment by vhvb1989** [comment]
@danieljurek @weshaggard , the provided SAS is not working :(  

```
Error: curl failed to put file to https://cppvcpkgcache.blob.core.windows.net/public-vcpkg-container/ec223fdb4f17e9881eae89fbc17131a8d19e51a344e6e2297da6f4699c5db8f9.zip?*** SECRET *** with exit code '0' and http code '403'
```

Can you help me with this?

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:22]: @vhvb1989 why do we need to do this here? Are you sure the SAS token has the correct value? I'd suggest you do any such transform inside the KV so you don't need to do any translations here.

Comment (vhvb1989) [reply eng/pipelines/templates/steps/vcpkg.yml:22]: Well, I was learning how to make this work during a Friday night w/o EngSys support. And I learned:

- When a variable is set as a secret, it needs to be mapped to any step/job which will need it.
- The SAS value (coming from EngSys contains a `?` as the first character. Probably copied from Azure Portal). And that's not useful here. So I needed to remove that character before using the value.
- After removing the character, the SAS has not write-access, so the `upload` cache operation failed.

At that point I gave up and added a comment about I need EngSys to provide a new SAS with write access.  And if the new SAS is provided without the `?` char at the start, then I can remove this logic to remove it.

**Comment by vhvb1989** [comment]
ping @weshaggard @danieljurek

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (danieljurek) [reply eng/pipelines/templates/steps/vcpkg.yml:22]: https://github.com/Azure/azure-sdk-for-cpp/issues/3439

File: eng/scripts/Set-VcpkgWriteModeCache.ps1
Comment (danieljurek) [inline eng/scripts/Set-VcpkgWriteModeCache.ps1:4]: @weshaggard -- The ps module helper fails when attempting to install `Az.Storage` on windows agents. I think we should add some kind of `-AllowClobber` feature

Example failure: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1454990&view=logs&j=34680a5f-6ae3-5138-ac88-c2c786351f1e&t=2a1986a3-0bcc-56b5-60f7-9b7c19a7a9f7&l=72

```
Installing module Az.Storage with min version 4.3.0 from https://www.powershellgallery.com/api/v2
PackageManagement\Install-Package : The following commands are already available on this system:'Start-CopyAzureStorageBlob,Stop-CopyAzureStorageBlob'. This module 'Az.Storage' may override the existing commands. If you still want to install this module 'Az.Storage', use -AllowClobber parameter.
At C:\program files\powershell\7\Modules\PowerShellGet\PSModule.psm1:9711 char:34
+ … talledPackages = PackageManagement\Install-Package @PSBoundParameters
+                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Microsoft.PowerShel…lets.InstallPackage:InstallPackage) [Install-Package], Exception
+ FullyQualifiedErrorId : CommandAlreadyAvailable,Validate-ModuleCommandAlreadyAvailable,Microsoft.PowerShell.PackageManagement.Cmdlets.InstallPackage
##[debug]Exit code: 1
##[debug]Leaving Invoke-VstsTool.
##[error]PowerShell exited with code '1'.
```

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:24]: We probably also want to add `condition: and(succeeded(), eq(variables['System.TeamProject'], 'internal'))` but that can be a later cleanup.

File: eng/scripts/Set-VcpkgWriteModeCache.ps1
Comment (weshaggard) [inline eng/scripts/Set-VcpkgWriteModeCache.ps1:9]: Do we know how long the default token time is for? We should make sure we don't want to shorten it.

Comment (danieljurek) [reply eng/scripts/Set-VcpkgWriteModeCache.ps1:9]: https://github.com/Azure/azure-sdk-for-cpp/issues/3480

File: eng/pipelines/templates/steps/vcpkg.yml
Comment (danieljurek) [reply eng/pipelines/templates/steps/vcpkg.yml:24]: https://github.com/Azure/azure-sdk-for-cpp/issues/3480

**Pull request merged**
